core/src/main/scala/org/ensime/indexer/SearchService.scala

```diff
@@ -316,50 +383,53 @@ class IndexingQueueActor(searchService: SearchService) extends Actor with ActorL
 
   override def receive: Receive = {
     case IndexFile(f) =>
-      todo += f.getName.getURI -> f
+      f match {
+        case jar if jar.getName.getExtension == "jar" => todo.addBinding(jar, jar)
+        case classFile => todo.addBinding(searchService.getTopLevelClassFile(classFile), classFile)
+      }
       debounce()
 
     case Process if todo.isEmpty => // nothing to do
 
     case Process =>
-      val (batch, remaining) = todo.splitAt(500)
-      todo = remaining
-      if (remaining.nonEmpty)
+      val batch = todo.take(250)
+      batch.keys.foreach(todo.remove)
+      if (todo.nonEmpty)
         debounce()
 
       import ExecutionContext.Implicits.global
 
-      log.debug(s"Indexing ${batch.size} files")
+      log.debug(s"Indexing ${batch.size} groups of files")
 
       def retry(): Unit = {
-        batch.foreach(self !)
+        batch.valuesIterator.foreach(_.foreach(self ! IndexFile(_)))
       }
 
-      Future.sequence(batch.map {
-        case (url, f) =>
-          val filename = f.getName.getPath
+      batch.grouped(10).foreach(chunk => Future.sequence(chunk.map {
```

Comment:
this soooo needs to be rewritten to use monix / swave / csp. It's become a monster.